src/ModelContextProtocol.AspNetCore/Authentication/McpAuthenticationHandler.cs

```diff
@@ -0,0 +1,151 @@
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using ModelContextProtocol.Authentication;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+
+namespace ModelContextProtocol.AspNetCore.Authentication;
+
+/// <summary>
+/// Authentication handler for MCP protocol that adds resource metadata to challenge responses
+/// and handles resource metadata endpoint requests.
+/// </summary>
+public class McpAuthenticationHandler : AuthenticationHandler<McpAuthenticationOptions>, IAuthenticationRequestHandler
+{
+    private readonly IOptionsMonitor<McpAuthenticationOptions> _optionsMonitor;
+    private string _resourceMetadataPath;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="McpAuthenticationHandler"/> class.
+    /// </summary>
+    public McpAuthenticationHandler(
+        IOptionsMonitor<McpAuthenticationOptions> options,
+        ILoggerFactory logger,
+        UrlEncoder encoder)
+        : base(options, logger, encoder)
+    {
+        _optionsMonitor = options;
```

Comment:
I think this is why you needed the unusual configureOptions logic in the AddMcp AuthenticationBuilder extension method. OptionsMonitor is already a protected property on the base class, so we don't need this property.

But instead of using `OptionsMonitor.CurrentValue` which will retrieve the default options rather than the one for the given stream, we should use just `Options` which gets initialized with the options for the right scheme when the authentication middleware call InitializeAsync.

https://github.com/dotnet/aspnetcore/blob/6a96694cda568cc8ebec2a0b34e408842eb9805f/src/Security/Authentication/Core/src/AuthenticationHandler.cs#L155